File: src/main/java/info/idgst/config/MailConfig.java
package info.idgst.config;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.Environment;
import org.springframework.mail.javamail.JavaMailSenderImpl;

import java.util.Properties;

/**
 * @author Aliaksei Bahdanau
 */
@Configuration
public class MailConfig {

    @Autowired IdgstConfigReader idgstConfigReader;

    @Bean
    public JavaMailSenderImpl mailSender(Environment env) {
        JavaMailSenderImpl mailSender = new JavaMailSenderImpl();
        mailSender.setJavaMailProperties(getProperties(env));
        mailSender.setHost(env.getProperty("email.smtp.host"));
        mailSender.setUsername(env.getProperty("email.login"));
        mailSender.setPassword(env.getProperty("email.password"));
        mailSender.setDefaultEncoding("UTF-8");

        return mailSender;
    }

    private Properties getProperties(Environment env) {
        Properties properties = new Properties();
        properties.put("mail.smtp.starttls.enable", env.getProperty("email.smtp.ttlsEnable"));
        return properties;
    }
}

File: src/main/kotlin/info/idgst/config/mail/MailConfig.kt
package info.idgst.config.mail

import info.idgst.config.mail.provider.MailProviderFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import org.springframework.mail.javamail.JavaMailSenderImpl
import java.util.*

/**
 * Configuration for [JavaMailSenderImpl]
 *
 * @author Aliaksei Bahdanau
 */
@Configuration
class MailConfig @Autowired constructor(private val mailProviderFactory: MailProviderFactory) {

    @Bean
    fun mailSender(env: Environment): JavaMailSenderImpl {
        val mailSender = JavaMailSenderImpl()
        mailSender.javaMailProperties = getProperties(env)
        mailSender.host = env.getProperty("email.smtp.host")
        mailSender.username = env.getProperty("email.login")
        mailSender.password = env.getProperty("email.password")
        mailSender.defaultEncoding = "UTF-8"

        return mailSender
    }

    private fun getProperties(env: Environment): Properties {
        val mailProvider = env.getProperty("email.provider")
        val mailPropertiesProvider = mailProviderFactory.getProvider(mailProvider)
        return mailPropertiesProvider.getMailProperties()
    }
}
